pkg/git_provider/gitlab.go

```diff
@@ -0,0 +1,356 @@
+			if err != nil {
+				return nil, err
+			}
+			if resp.StatusCode != 201 {
+				return nil, fmt.Errorf("failed to create repo level webhhok for %s, API returned %d", *repo, resp.StatusCode)
+			}
+			log.Printf("created webhook for %s: %s\n", *repo, gitlabHook.URL)
+		} else {
+			
+			editProjectHookOpts := gitlab.EditProjectHookOptions{
+				URL: gitlab.Ptr(c.cfg.GitProviderConfig.WebhookURL),
+				Token: gitlab.Ptr(c.cfg.GitProviderConfig.WebhookSecret),
+				MergeRequestsEvents: gitlab.Ptr(true),
+				PushEvents: gitlab.Ptr(true),
+				ReleasesEvents: gitlab.Ptr(true),
+				TagPushEvents: gitlab.Ptr(true),
+			}
+			gitlabHook, resp, err := c.client.Projects.EditProjectHook(repo, respHook.ID, &editProjectHookOpts)
+			if err != nil {
+				return nil, err
+			}
+			if resp.StatusCode != http.StatusOK {
+				return nil, fmt.Errorf("failed to update repo level webhhok for %s, API returned %d", *repo, resp.StatusCode)
+			}
+			log.Printf("edited webhook for %s: %s\n", *repo, gitlabHook.URL)
+		}
+
+	}
+
+	hookID := int64(gitlabHook.ID)
+	return &HookWithStatus{HookID: hookID, HealthStatus: true, RepoName: repo}, nil
+}
+
+func (c *GitlabClientImpl) UnsetWebhook(ctx *context.Context, hook *HookWithStatus) error {
+
+	if hook.RepoName == nil {
+		resp, err := c.client.Groups.DeleteGroupHook( c.cfg.GitProviderConfig.OrgName, int(hook.HookID))
+		if err != nil {
+			return err
+		}
+
+		if resp.StatusCode != 204 {
+			return fmt.Errorf("failed to delete group level webhhok, API call returned %d", resp.StatusCode)
+		}
+		log.Printf("removed group webhook, hookID :%d\n", hook.HookID)
+	} else {
+		resp, err := c.client.Projects.DeleteProjectHook(*hook.RepoName, int(hook.HookID))
+
+		if err != nil {
+			return fmt.Errorf("failed to delete project level webhhok for %s, API call returned %d. %s", *hook.RepoName, resp.StatusCode, err)
+		}
+
+		if resp.StatusCode != 204 {
+			return fmt.Errorf("failed to delete project level webhhok for %s, API call returned %d", *hook.RepoName, resp.StatusCode)
+		}
+		log.Printf("removed project webhook, project:%s hookID :%d\n", *hook.RepoName, hook.HookID) // INFO
+	}
+
+	return nil
+}
+
+func (c *GitlabClientImpl) HandlePayload(ctx *context.Context, request *http.Request, secret []byte) (*WebhookPayload, error) {
+	var webhookPayload *WebhookPayload
+
+	payload, err := validatePayload(request, secret)
+	if err != nil {
+		return nil, err
+	}
+
+	event, err := gitlab.ParseWebhook(gitlab.WebhookEventType(request), payload)
+	if err != nil {
+		return nil, err
+	}
+
+
+	switch e := event.(type) {
+	case gitlab.PushEvent:
+		webhookPayload = &WebhookPayload{
+			Event:     "push",
+			Action:    e.EventName,
+			Repo:      e.Project.Name,
+			Branch:    strings.TrimPrefix(e.Ref, "refs/heads/"),
+			Commit:    e.CheckoutSHA,
+			User:      e.UserName,
+			UserEmail: e.UserEmail,
+			OwnerID:   int64(e.UserID),
+		}
+	case gitlab.MergeEvent:
+		webhookPayload = &WebhookPayload{
+			Event:            "pull_request",
```

Comment:
it is merge_request
https://docs.gitlab.com/ee/user/project/integrations/webhook_events.html#merge-request-events